node/pkg/p2p/watermark_test.go

```diff
@@ -0,0 +1,192 @@
+	priv                   p2pcrypto.PrivKey
+	gk                     *ecdsa.PrivateKey
+	gst                    *node_common.GuardianSetState
+	networkID              string
+	bootstrapPeers         string
+	nodeName               string
+	disableHeartbeatVerify bool
+	rootCtxCancel          context.CancelFunc
+	gov                    *governor.ChainGovernor
+	acct                   *accountant.Accountant
+	signedGovCfg           chan *gossipv1.SignedChainGovernorConfig
+	signedGovSt            chan *gossipv1.SignedChainGovernorStatus
+	components             *Components
+}
+
+func NewG(nodeName string) *G {
+	cs := 20
+	p2ppriv, _, err := p2pcrypto.GenerateKeyPair(p2pcrypto.Ed25519, -1)
+	if err != nil {
+		panic(err)
+	}
+
+	guardianpriv, err := ecdsa.GenerateKey(crypto.S256(), rand.Reader)
+	if err != nil {
+		panic(err)
+	}
+
+	g := &G{
+		obsvC:                  make(chan *gossipv1.SignedObservation, cs),
+		obsvReqC:               make(chan *gossipv1.ObservationRequest, cs),
+		obsvReqSendC:           make(chan *gossipv1.ObservationRequest, cs),
+		sendC:                  make(chan []byte, cs),
+		signedInC:              make(chan *gossipv1.SignedVAAWithQuorum, cs),
+		priv:                   p2ppriv,
+		gk:                     guardianpriv,
+		gst:                    node_common.NewGuardianSetState(nil),
+		nodeName:               nodeName,
+		disableHeartbeatVerify: false,
+		rootCtxCancel:          nil,
+		gov:                    nil,
+		signedGovCfg:           make(chan *gossipv1.SignedChainGovernorConfig, cs),
+		signedGovSt:            make(chan *gossipv1.SignedChainGovernorStatus, cs),
+		components:             DefaultComponents(),
+	}
+
+	// Consume all output channels
+	go func() {
+		name := g.nodeName
+		fmt.Printf("[%s] consuming\n", name)
+		select {
+		case <-g.obsvC:
+			fmt.Printf("[%s] g.obsvC\n", name)
+		case <-g.obsvReqC:
+			fmt.Printf("[%s] g.obsvReqC\n", name)
+		case <-g.signedInC:
+			fmt.Printf("[%s] g.signedInC\n", name)
+		case <-g.signedGovCfg:
+			fmt.Printf("[%s] g.signedGovCfg\n", name)
+		case <-g.signedGovSt:
+			fmt.Printf("[%s] g.signedGovSt\n", name)
+		case <-g.sendC:
+			fmt.Printf("[%s] g.sendC\n", name)
+		}
+	}()
+
+	return g
+}
+
+// TestWatermark runs 4 different guardians one of which does not send its P2PID in the signed part of the heartbeat.
+// The expectation is that hosts that send this information will become "protected" by the Connection Manager.
+func TestWatermark(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Create 4 nodes
+	var guardianset = &node_common.GuardianSet{}
+	var gs [4]*G
+	for i, _ := range gs {
+		gs[i] = NewG(fmt.Sprintf("n%d", i))
+		gs[i].components.Port = uint(11000 + i)
+		gs[i].networkID = "/wormhole/localdev"
+
+		guardianset.Keys = append(guardianset.Keys, crypto.PubkeyToAddress(gs[i].gk.PublicKey))
+
+		id, err := p2ppeer.IDFromPublicKey(gs[0].priv.GetPublic())
+		if err != nil {
+			panic(err)
+		}
```

Comment:
```suggestion
require.NotError(err)
```

Panicking in a unit test is not nice. Panics should be used for safety critical parts of the real code. In unit tests, you can panic or just make a false assertion so the test gracefully and nicely fails.

Using [require](https://pkg.go.dev/github.com/stretchr/testify/require) will stop further execution of this test and mark it nicely as failed vs spew a big panic with the full stack of every single goroutine to the console.